develop/migration_mt_django.rst

```diff
@@ -5,76 +5,185 @@
+- Changing the primary keys of distributed tables to include the tenant column
+- Updating the models to use the :code:`TenantModelMixin`
+- Distributing the data
+- Updating the Django Application to scope queries
+
 
 Preparing to scale-out a multi-tenant application
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Initially you’ll start with all tenants placed on a single database node. Django's typical conventions make some assumptions about the data storage which limit scale-out options.
+Initially you’ll start with all tenants placed on a single database node. To be able to scale out django, some simple changes will have to be made to your models.
 
-In particular, the ORM introduces a pattern where you normalize data and split it into many distinct models each identified by a single ``id`` column (usually added implicitly by the ORM). For instance, consider this simplified model:
+Let's consider this simplified model:
 
 .. code-block:: python
 
   from django.utils import timezone
   from django.db import models
 
-  class Store(models.Model):
-    name = models.CharField(max_length=255)
-    url = models.URLField()
+  class Country(models.Model):
+      name = models.CharField(max_length=255)
+
+  class Account(models.Model):
+      name = models.CharField(max_length=255)
+      domain = models.CharField(max_length=255)
+      subdomain = models.CharField(max_length=255)
+      country = models.ForeignKey(Country, on_delete=models.CASCADE)
+
+
+  class Manager(models.Model):
+      name = models.CharField(max_length=255)
+      account = models.ForeignKey(Account, on_delete=models.CASCADE,
+                                  related_name='managers')
+
+
+  class Project(models.Model):
+      name = models.CharField(max_length=255)
+      account = models.ForeignKey(Account, related_name='projects',
+                                  on_delete=models.CASCADE)
+      managers = models.ManyToManyField(Manager)
+
+
+  class Task(models.Model):
+      name = models.CharField(max_length=255)
+      project = models.ForeignKey(Project, on_delete=models.CASCADE,
+                                  related_name='tasks')
+
+
+
+The tricky thing with this pattern is that in order to find all tasks for an account, you'll have to query for all of a account's project first. This becomes a problem once you start sharding data, and in particular when you run UPDATE or DELETE queries on nested models like task in this example.
+
+
+1. Introducing the tenant column to models belonging to an account
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+
+**1.1 Introducting the column to models belonging to an account**
+
+In order to scale out a multi-tenant model, it's essential that you can locate all records that belong to an account quickly.
+The reason is that we want that an ORM call like
+
+.. code-block:: python
+
+  Project.objects.filter(account_id=1).prefetch_related('tasks')
+
+Generating the following queries
```

Comment:
The before and after SQL kind of confused me the way it was phrased. What about this.

```rst
In order to scale out a multi-tenant model, it’s essential for queries to quickly
locate all records that belong to an account. Consider an ORM call such as:

.. code-block:: python

  Project.objects.filter(account_id=1).prefetch_related('tasks')

It generates these underlying SQL queries:

.. code-block:: sql

  SELECT *
    FROM myapp_project
   WHERE account_id = 1;
  
  SELECT *
    FROM myapp_task
   WHERE project_id IN (1, 2, 3);

However the second query would go faster with an extra filter:

.. code-block:: diff

  + -- the AND clause identifies the tenant

    SELECT *
      FROM myapp_task
     WHERE project_id IN (1, 2, 3)
  +    AND account_id = 1;
```